File: tests/unit/xmltv-parser.test.ts
import { describe, it, expect, vi } from 'vitest';

// Mock electron-log before importing the module
vi.mock('electron-log/main', () => ({
  default: { info: vi.fn(), warn: vi.fn(), error: vi.fn() },
}));

import { parseXmltv, parseXmltvTimestamp } from '../../src/main/services/xmltv-parser';

describe('XMLTV Parser', () => {
  describe('parseXmltvTimestamp', () => {
    it('parses timestamp without offset (UTC)', () => {
      const ts = parseXmltvTimestamp('20260415120000');
      // 2026-04-15T12:00:00Z
      expect(ts).toBe(Math.floor(new Date('2026-04-15T12:00:00Z').getTime() / 1000));
    });

    it('parses timestamp with positive offset', () => {
      const ts = parseXmltvTimestamp('20260415120000 +0200');
      // 2026-04-15T12:00:00+02:00 = 2026-04-15T10:00:00Z
      expect(ts).toBe(Math.floor(new Date('2026-04-15T10:00:00Z').getTime() / 1000));
    });

    it('parses timestamp with negative offset', () => {
      const ts = parseXmltvTimestamp('20260415120000 -0500');
      // 2026-04-15T12:00:00-05:00 = 2026-04-15T17:00:00Z
      expect(ts).toBe(Math.floor(new Date('2026-04-15T17:00:00Z').getTime() / 1000));
    });

    it('handles timestamp with no spaces before offset', () => {
      const ts = parseXmltvTimestamp('20260101000000+0000');
      expect(ts).toBe(Math.floor(new Date('2026-01-01T00:00:00Z').getTime() / 1000));
    });

    it('returns 0 for empty string', () => {
      expect(parseXmltvTimestamp('')).toBe(0);
    });

    it('returns 0 for invalid format', () => {
      expect(parseXmltvTimestamp('not-a-timestamp')).toBe(0);
      expect(parseXmltvTimestamp('2026')).toBe(0);
      expect(parseXmltvTimestamp('20261301000000')).toBe(0); // month 13 → invalid date
    });

    it('trims whitespace', () => {
      const ts = parseXmltvTimestamp('  20260415120000 +0000  ');
      expect(ts).toBe(Math.floor(new Date('2026-04-15T12:00:00Z').getTime() / 1000));
    });
  });

  describe('parseXmltv', () => {
    it('parses channels from XMLTV', async () => {
      const xml = `<?xml version="1.0"?>
<tv>
  <channel id="bbc1">
    <display-name>BBC One</display-name>
    <icon src="http://logos.com/bbc1.png" />
  </channel>
  <channel id="itv1">
    <display-name>ITV</display-name>
  </channel>
</tv>`;

      const result = await parseXmltv(xml);
      expect(result.channels).toHaveLength(2);
      expect(result.channels[0]).toEqual({
        id: 'bbc1',
        displayName: 'BBC One',
        iconUrl: 'http://logos.com/bbc1.png',
      });
      expect(result.channels[1]).toEqual({
        id: 'itv1',
        displayName: 'ITV',
        iconUrl: undefined,
      });
    });

    it('parses programmes from XMLTV', async () => {
      const xml = `<?xml version="1.0"?>
<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="bbc1">
    <title>Evening News</title>
    <desc>The latest headlines.</desc>
    <category>News</category>
    <icon src="http://img.com/news.png" />
  </programme>
  <programme start="20260415190000 +0000" stop="20260415200000 +0000" channel="bbc1">
    <title>Drama Show</title>
  </programme>
</tv>`;

      const result = await parseXmltv(xml);
      expect(result.programmes).toHaveLength(2);

      const p1 = result.programmes[0];
      expect(p1.channelId).toBe('bbc1');
      expect(p1.title).toBe('Evening News');
      expect(p1.description).toBe('The latest headlines.');
      expect(p1.category).toBe('News');
      expect(p1.iconUrl).toBe('http://img.com/news.png');
      expect(p1.startTime).toBe(parseXmltvTimestamp('20260415180000 +0000'));
      expect(p1.endTime).toBe(parseXmltvTimestamp('20260415190000 +0000'));

      const p2 = result.programmes[1];
      expect(p2.title).toBe('Drama Show');
      expect(p2.description).toBeUndefined();
      expect(p2.category).toBeUndefined();
    });

    it('skips programmes missing required attributes', async () => {
      const xml = `<tv>
  <programme start="20260415180000 +0000" channel="bbc1">
    <title>Missing stop attr</title>
  </programme>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="bbc1">
    <title>Valid</title>
  </programme>
</tv>`;

      const result = await parseXmltv(xml);
      expect(result.programmes).toHaveLength(1);
      expect(result.programmes[0].title).toBe('Valid');
    });

    it('skips programmes with no title', async () => {
      const xml = `<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="bbc1">
    <desc>No title element</desc>
  </programme>
</tv>`;

      const result = await parseXmltv(xml);
      expect(result.programmes).toHaveLength(0);
    });

    it('decodes XML entities in text', async () => {
      const xml = `<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="ch1">
    <title>Tom &amp; Jerry</title>
    <desc>It&apos;s a &quot;classic&quot;</desc>
  </programme>
</tv>`;

      const result = await parseXmltv(xml);
      expect(result.programmes[0].title).toBe('Tom & Jerry');
      expect(result.programmes[0].description).toBe('It\'s a "classic"');
    });

    it('handles empty XML gracefully', async () => {
      const result = await parseXmltv('<tv></tv>');
      expect(result.channels).toEqual([]);
      expect(result.programmes).toEqual([]);
    });

    it('handles plain UTF-8 buffer input', async () => {
      const xml = `<tv>
  <channel id="ch1"><display-name>Test</display-name></channel>
</tv>`;
      const buf = Buffer.from(xml, 'utf-8');
      const result = await parseXmltv(buf);
      expect(result.channels).toHaveLength(1);
      expect(result.channels[0].id).toBe('ch1');
    });

    it('handles localized title elements with lang attribute', async () => {
      const xml = `<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="ch1">
    <title lang="en">Hello World</title>
  </programme>
</tv>`;

      const result = await parseXmltv(xml);
      expect(result.programmes[0].title).toBe('Hello World');
    });

    it('handles numeric character references', async () => {
      const xml = `<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="ch1">
    <title>&#72;ello</title>
  </programme>
</tv>`;

      const result = await parseXmltv(xml);
      expect(result.programmes[0].title).toBe('Hello');
    });
  });
});

File: packages/shared/src/commonTest/kotlin/com/yancotv/shared/parsers/XmltvParserTest.kt
package com.yancotv.shared.parsers

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class XmltvParserTest {
    // --- parseXmltvTimestamp ---

    @Test
    fun parsesTimestampWithoutOffsetUtc() {
        // 2026-04-15T12:00:00Z — precomputed:
        // 2026-01-01T00:00:00Z = 1767225600; +104d +12h = 1776254400
        val ts = parseXmltvTimestamp("20260415120000")
        assertEquals(1776254400L, ts)
    }

    @Test
    fun parsesTimestampWithPositiveOffset() {
        // 2026-04-15T12:00:00+02:00 = 2026-04-15T10:00:00Z = 1776254400 - 2h
        val ts = parseXmltvTimestamp("20260415120000 +0200")
        assertEquals(1776254400L - 2 * 3600L, ts)
    }

    @Test
    fun parsesTimestampWithNegativeOffset() {
        // 2026-04-15T12:00:00-05:00 = 2026-04-15T17:00:00Z = 1776254400 + 5h
        val ts = parseXmltvTimestamp("20260415120000 -0500")
        assertEquals(1776254400L + 5 * 3600L, ts)
    }

    @Test
    fun handlesTimestampWithNoSpacesBeforeOffset() {
        // 2026-01-01T00:00:00Z = 1767225600
        val ts = parseXmltvTimestamp("20260101000000+0000")
        assertEquals(1767225600L, ts)
    }

    @Test
    fun returnsZeroForEmptyString() {
        assertEquals(0L, parseXmltvTimestamp(""))
    }

    @Test
    fun returnsZeroForInvalidFormat() {
        assertEquals(0L, parseXmltvTimestamp("not-a-timestamp"))
        assertEquals(0L, parseXmltvTimestamp("2026"))
        assertEquals(0L, parseXmltvTimestamp("20261301000000")) // month 13
    }

    @Test
    fun trimsWhitespace() {
        val ts = parseXmltvTimestamp("  20260415120000 +0000  ")
        assertEquals(1776254400L, ts)
    }

    // --- parseXmltv ---

    @Test
    fun parsesChannelsFromXmltv() {
        val xml = """<?xml version="1.0"?>
<tv>
  <channel id="bbc1">
    <display-name>BBC One</display-name>
    <icon src="http://logos.com/bbc1.png" />
  </channel>
  <channel id="itv1">
    <display-name>ITV</display-name>
  </channel>
</tv>"""

        val result = parseXmltv(xml)
        assertEquals(2, result.channels.size)
        assertEquals(
            XmltvChannel(id = "bbc1", displayName = "BBC One", iconUrl = "http://logos.com/bbc1.png"),
            result.channels[0],
        )
        assertEquals(
            XmltvChannel(id = "itv1", displayName = "ITV", iconUrl = null),
            result.channels[1],
        )
    }

    @Test
    fun parsesProgrammesFromXmltv() {
        val xml = """<?xml version="1.0"?>
<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="bbc1">
    <title>Evening News</title>
    <desc>The latest headlines.</desc>
    <category>News</category>
    <icon src="http://img.com/news.png" />
  </programme>
  <programme start="20260415190000 +0000" stop="20260415200000 +0000" channel="bbc1">
    <title>Drama Show</title>
  </programme>
</tv>"""

        val result = parseXmltv(xml)
        assertEquals(2, result.programmes.size)

        val p1 = result.programmes[0]
        assertEquals("bbc1", p1.channelId)
        assertEquals("Evening News", p1.title)
        assertEquals("The latest headlines.", p1.description)
        assertEquals("News", p1.category)
        assertEquals("http://img.com/news.png", p1.iconUrl)
        assertEquals(parseXmltvTimestamp("20260415180000 +0000"), p1.startTime)
        assertEquals(parseXmltvTimestamp("20260415190000 +0000"), p1.endTime)

        val p2 = result.programmes[1]
        assertEquals("Drama Show", p2.title)
        assertNull(p2.description)
        assertNull(p2.category)
    }

    @Test
    fun skipsProgrammesMissingRequiredAttributes() {
        val xml = """<tv>
  <programme start="20260415180000 +0000" channel="bbc1">
    <title>Missing stop attr</title>
  </programme>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="bbc1">
    <title>Valid</title>
  </programme>
</tv>"""

        val result = parseXmltv(xml)
        assertEquals(1, result.programmes.size)
        assertEquals("Valid", result.programmes[0].title)
    }

    @Test
    fun skipsProgrammesWithNoTitle() {
        val xml = """<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="bbc1">
    <desc>No title element</desc>
  </programme>
</tv>"""

        val result = parseXmltv(xml)
        assertEquals(0, result.programmes.size)
    }

    @Test
    fun decodesXmlEntitiesInText() {
        val xml = """<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="ch1">
    <title>Tom &amp; Jerry</title>
    <desc>It&apos;s a &quot;classic&quot;</desc>
  </programme>
</tv>"""

        val result = parseXmltv(xml)
        assertEquals("Tom & Jerry", result.programmes[0].title)
        assertEquals("It's a \"classic\"", result.programmes[0].description)
    }

    @Test
    fun handlesEmptyXmlGracefully() {
        val result = parseXmltv("<tv></tv>")
        assertTrue(result.channels.isEmpty())
        assertTrue(result.programmes.isEmpty())
    }

    @Test
    fun handlesPlainUtf8BufferInput() {
        val xml = """<tv>
  <channel id="ch1"><display-name>Test</display-name></channel>
</tv>"""
        val buf = xml.encodeToByteArray()
        val result = parseXmltv(buf)
        assertEquals(1, result.channels.size)
        assertEquals("ch1", result.channels[0].id)
    }

    @Test
    fun handlesLocalizedTitleElementsWithLangAttribute() {
        val xml = """<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="ch1">
    <title lang="en">Hello World</title>
  </programme>
</tv>"""

        val result = parseXmltv(xml)
        assertEquals("Hello World", result.programmes[0].title)
    }

    @Test
    fun handlesNumericCharacterReferences() {
        val xml = """<tv>
  <programme start="20260415180000 +0000" stop="20260415190000 +0000" channel="ch1">
    <title>&#72;ello</title>
  </programme>
</tv>"""

        val result = parseXmltv(xml)
        assertEquals("Hello", result.programmes[0].title)
    }
}
